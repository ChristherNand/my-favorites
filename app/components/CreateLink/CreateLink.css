.createLink {
  outline: none;
  text-align: center;
  align-items: center;
  box-sizing: border-box;
}

.createLink > *:not(h2) {
  margin-top: 10px;
}

.createLink .createLink__contInputs {
}

.createLink__contInputs__title {
}

.createLink__contTags {
  width: 100%;
}

.createLink__contTags .autoComplete {
  width: 200px;
  margin-right: 10px;
}