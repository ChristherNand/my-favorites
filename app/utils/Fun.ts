export default {
  noon: (): null => null,
};
