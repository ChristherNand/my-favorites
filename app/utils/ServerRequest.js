import instance from './axios.conf';
import {ERROR_TYPES, ERROR_MESSAGES} from "../contans";

export function getTags(){
  return instance.get('/tag')
          .then(({data}) => data)
          .catch(catchError);
}

export function createTag(info){
  return instance.put('/tag', info)
          .then(({data}) => data)
          .catch(catchError);
      }
        
export function updateTagRequest(id, name){
  return instance.put(`/tag/${id}`, {name})
          .then(({data}) => data)
          .catch(catchError);
}

export function deleteTag(id){
  return instance.delete(`/tag/${id}`)
          .then(({data}) => data)
          .catch(catchError);
}

export function getLinks(){
  return instance.get('/link')
          .then(({data}) => data)
          .catch(catchError);
}

export function getLinksByTags(tags){
  return instance.get(`/link/byTags/${tags}`)
          .then(({data}) => data)
          .catch(catchError);
}

export function createLink(info){
  return instance.put('/link', info)
          .then(({data}) => data)
          .catch(catchError);
}

export function removeLinkRequest(id){
  return instance.delete(`/link/${id}`)
          .then(({data}) => data)
          .catch(catchError);
}

export function removeTagLinkRequest(linkID, tagID){
  return instance.put(`/link/${linkID}/${tagID}`)
          .then(({data}) => data)
          .catch(catchError);
}

function catchError(error){
  console.log('---- my handle error ---');
  switch(true){
    case error.message.toLocaleLowerCase().includes('network'):
      throw({type: ERROR_TYPES.NETWORK, info: ERROR_MESSAGES.NETWORK_ERROR});
    default: 
      throw(error);
  }

}