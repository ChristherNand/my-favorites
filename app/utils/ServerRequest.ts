import { iLink, iTag } from '../interfaces';
import { iNewLink } from '../interfaces/iLink';
import instance from './axios.conf';
import catchError from './CatchError';

type appPromise<T> = Promise<{ status: string; data: T }>;

export function getTags(): Promise<iTag[]> {
  return instance
    .get('/tag')
    .then(({ data }) => data)
    .catch(catchError);
}
// Todo add return types
export function createTag(info: iTag) {
  return instance
    .put('/tag', info)
    .then(({ data }) => data)
    .catch(catchError);
}

export function updateTagRequest(id, name) {
  return instance
    .put(`/tag/${id}`, { name })
    .then(({ data }) => data)
    .catch(catchError);
}

export function deleteTag(id: string) {
  return instance
    .delete(`/tag/${id}`)
    .then(({ data }) => data)
    .catch(catchError);
}

export function getLinks() {
  return instance
    .get('/link')
    .then(({ data }) => data)
    .catch(catchError);
}

export function getLinksByTags(tags: string) {
  return instance
    .get(`/link/byTags/${tags}`)
    .then(({ data }) => data)
    .catch(catchError);
}

export function createLink(info: iNewLink): appPromise<iLink> {
  return instance
    .put('/link', info)
    .then(({ data }) => data)
    .catch(catchError);
}

export function removeLinkRequest(id) {
  return instance
    .delete(`/link/${id}`)
    .then(({ data }) => data)
    .catch(catchError);
}

export function removeTagLinkRequest(linkID, tagID) {
  return instance
    .put(`/link/${linkID}/${tagID}`)
    .then(({ data }) => data)
    .catch(catchError);
}
