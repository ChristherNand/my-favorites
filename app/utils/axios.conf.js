const axios = require('axios');
const port = 8082;

const config = {
  port,
  url: process.env.API_URL || `http://localhost:${port}/api`
}

const instance = axios.create({
  baseURL: config.url,
  timeout: 1000,
});

instance.interceptors.request.use(function (config) {
  console.log("Sending request...")
  return config;
}, function (error) {
  return Promise.reject(error);
});

// Add a response interceptor
instance.interceptors.response.use(function (response) {
  console.log("getting request...")
  return response;
}, function (error) {
  return Promise.reject(error);
});

module.exports = instance;
