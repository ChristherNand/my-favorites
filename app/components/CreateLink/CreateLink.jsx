import React, {useRef, useState, useEffect} from 'react';
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {TagPropType} from "../../propsTypes";
import "./CreateLink.css"

import Tag from "../generals/Tag";
import AutoComplete from "../generals/AutoComplete";
import {addLink, invalidLink, getAllTags} from "../../state/actions";

const mapStateToProps = state => {
  return {
    tags: state.tags,
    isInvalidLink: state.invalidLink
  }
}

const mapDispachToProps = dispatch => {
  return {
    addLink: info => dispatch(addLink(info)),
    invalidLink: () => dispatch(invalidLink()),
    getAllTags: () => dispatch(getAllTags()),
  }
}

export function CreateLink(props) {
  const inputTitle = useRef();
  const inputUrl = useRef();
  const [tags, setTags] = useState([]);
  console.log('CreateLink props', props);

  useEffect(() => {
    if(!props.tags.length){
      props.getAllTags();
    }
  }, []);

  function removeInvalid(event){
    event.target.classList.remove("invalid");
  }

  function check() {
    let isValid = true;
    if(inputTitle.current.value === "") {
        inputTitle.current.classList.add("invalid")
        isValid = false;
    }
    
    if(inputUrl.current.value === "") {
        inputUrl.current.classList.add("invalid")
        isValid = false;
    }

    if(isValid){
        props.addLink({
            title: inputTitle.current.value,
            url: inputUrl.current.value, 
            tags: tags.map(tag => tag.id)
        });
        clear();
    }
  }

  function clear(){
    inputUrl.current.value = "";
    inputTitle.current.value = "";
    setTags([]);
  }

  return  <div className="createLink">
            <h2>Create Link</h2>

            <input className="createLink__contInputs__title" placeholder="Title" onFocus={removeInvalid} ref={inputTitle}/>
            <input className="createLink__contInputs__url" placeholder="Url" onFocus={removeInvalid} ref={inputUrl}/>
            <button className="createLink__send" type="button" onClick={check}>Send</button>

            <div className="createLink__contTags --flex">
              <AutoComplete 
                  autoHide={false}
                  propertyFilter="name" 
                  options={props.tags} 
                  placeHolder="Add Tags"
                  clearAfterSelecting={true}
                  onSelected={tag => setTags([{...tag}, ...tags])}/>

              <div className="contOptions">
                  {tags.map((tag, index) => <Tag key={`${index}-boardTags`} {...tag}/>)}
              </div>
            </div>
          </div>
}

CreateLink.propTypes = {
  addLink: PropTypes.func,
  tags: PropTypes.arrayOf(PropTypes.shape(TagPropType))
}

export default connect(mapStateToProps, mapDispachToProps)(CreateLink);
