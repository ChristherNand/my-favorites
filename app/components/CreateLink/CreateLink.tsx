import React, { useRef, useState, useEffect, Dispatch } from 'react';

import './CreateLink.css';

import TagList from '../generals/TagList';
import {
  addLinkAsyncAction,
  getAllTagsAsyncAction,
  setLsUrlAction,
  setLsTitleAction,
  setLsTagsAction,
  clearLsAction,
} from '../../state/actions';
import iCreateLink from '../../interfaces/iCreateLink';
import { RootState } from '../../state/store';
import { useAppDispatch, useAppSelector } from '../../hooks/redux';
import { iNewLink } from '../../interfaces/iLink';
import { iTag } from '../../interfaces';

const selectTags = (state: RootState) => state.tags;
const selectLocalStorage = (state: RootState) => state.localStorage;
//todo Implement invalid toast
// const selectInvalidLink = (state: RootState) => state.validation.invalidLink;

const CreateLink: React.FC<iCreateLink> = () => {
  const inputTitle = useRef<HTMLInputElement>(null);
  const inputUrl = useRef<HTMLInputElement>(null);
  const [clearList, setClearList] = useState(false);

  const tags = useAppSelector(selectTags);
  const lsLink = useAppSelector(selectLocalStorage);
  //todo Implement invalid toast
  // const isInvalidLink = useAppSelector(selectInvalidLink);
  const dispatchAddLink: Dispatch<any> = useAppDispatch();
  const dispatchGetAllTags: Dispatch<any> = useAppDispatch();
  const dispatchClearLs = useAppDispatch();
  const dispatchSetLsUrl = useAppDispatch();
  const dispatchSetLsTitle = useAppDispatch();
  const dispatchSetLsTags = useAppDispatch();

  const addLink = (link: iNewLink) => dispatchAddLink(addLinkAsyncAction(link));
  const getAllTags = () => dispatchGetAllTags(getAllTagsAsyncAction());
  const clearLs = () => dispatchClearLs(clearLsAction());
  const setLsUrl = (url: string) => dispatchSetLsUrl(setLsUrlAction(url));
  const setLsTitle = (url: string) => dispatchSetLsTitle(setLsTitleAction(url));
  const setLsTags = (tags: iTag[]) => dispatchSetLsTags(setLsTagsAction(tags));

  useEffect(() => {
    if (!tags?.length) {
      getAllTags();
    }

    inputUrl.current.value = lsLink?.url;
    inputTitle.current.value = lsLink?.title;
  }, []);

  const removeInvalid = (event) => event?.target?.classList?.remove('invalid');

  //Todo, the 'undefined' validation was added to make the test pass, find a better approach.
  const isInvalidInput = (text: string) => !text || text === 'undefined';

  const check = () => {
    let isValid = true;

    if (isInvalidInput(inputTitle?.current?.value)) {
      inputTitle.current.classList.add('invalid');
      isValid = false;
    }

    if (isInvalidInput(inputUrl?.current?.value)) {
      inputUrl.current.classList.add('invalid');
      isValid = false;
    }

    if (isValid) {
      addLink({
        title: inputTitle.current.value,
        url: inputUrl.current.value,
        tags: lsLink?.tags?.map((tag) => tag.id as string),
      });
      clear();
    }
  };

  const clear = () => {
    inputUrl.current.value = '';
    inputTitle.current.value = '';
    setClearList(true);
    clearLs();
  };

  const updateTitle = () => setLsTitle(inputTitle?.current?.value);
  const updateUrl = () => setLsUrl(inputUrl?.current?.value);

  return (
    <div className="createLink">
      <h2>Create Link</h2>

      <input
        placeholder="Title"
        ref={inputTitle}
        data-test="inp-title"
        onChange={updateTitle}
        onFocus={removeInvalid}
        className="createLink__contInputs__title"
      />
      <div className="--flex">
        <input
          placeholder="Url"
          ref={inputUrl}
          data-test="inp-url"
          onChange={updateUrl}
          onFocus={removeInvalid}
          className="createLink__contInputs__url"
        />
        <button
          className="createLink__send button"
          type="button"
          onClick={check}
          data-test="btn-send"
        >
          Send
        </button>
      </div>

      <TagList
        className="createLink__contTags --flex"
        options={tags}
        autoHide={false}
        clearList={clearList}
        data-test="cp-tagList"
        placeHolder="Add Tags"
        onTagsSaved={setLsTags}
        clearAfterSelecting={true}
        initialSavedTags={lsLink?.tags}
      />
    </div>
  );
};

export default CreateLink;
