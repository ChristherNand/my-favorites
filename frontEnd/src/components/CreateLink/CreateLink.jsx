import React, {useRef, useState} from 'react';
import {connect} from "react-redux";
import "./CreateLink.css"

import Tag from "../generals/Tag";
import AutoComplete from "../generals/AutoComplete";
import {addLink, invalidLink} from "../../state/actions";

const mapStateToProps = state => {
    return {
        tags: state.tags,
        isInvalidLink: state.invalidLink
    }
}

const mapDispachToProps = dispatch => {
    return {
        addLink: info => dispatch(addLink(info)),
        invalidLink: () => dispatch(invalidLink()),
    }
}

function CreateLink(props) {
    const inputTitle = useRef();
    const inputUrl = useRef();
    const [tags, setTags] = useState([]);

    function removeInvalid(event){
        event.target.classList.remove("invalid");
    }

    function check() {
        let isValid = true;
        if(inputTitle.current.value === "") {
            inputTitle.current.classList.add("invalid")
            isValid = false;
        }
        
        if(inputUrl.current.value === "") {
            inputUrl.current.classList.add("invalid")
            isValid = false;
        }

        if(isValid){
            props.addLink({
                title: inputTitle.current.value,
                url: inputUrl.current.value, 
                tags: tags.map(tag => tag.id)
            });
            clear();
        }
    }

    function clear(){
        inputUrl.current.value = "";
        inputTitle.current.value = "";
        setTags([]);
    }

    return  <div className="createLink --flex --wrap">
                <div className="createLink__contInputs --flex">
                    <input placeholder="Title" onFocus={removeInvalid} ref={inputTitle}/>
                    <input placeholder="Url" onFocus={removeInvalid} ref={inputUrl}/>
                </div>
                <button type="button" onClick={check}>Send</button>

                <div className="createLink__contTags --flex">
                    <AutoComplete 
                        autoHide={false}
                        propertyFilter="name" 
                        options={props.tags} 
                        placeHolder="Add Tags"
                        clearAfterSelecting={true}
                        onSelected={tag => setTags([{...tag}, ...tags])}/>

                    <div className="contOptions">
                        {tags.map((tag, index) => <Tag key={`${index}-boardTags`} {...tag}/>)}
                    </div>
                </div>
            </div>
}

export default connect(mapStateToProps, mapDispachToProps)(CreateLink);
