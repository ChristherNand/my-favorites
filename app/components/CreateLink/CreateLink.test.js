import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { CreateLink } from './CreateLink';
import { setPropTypes, findByTestAttr } from '../../utils/test/';

var Component;
const initialProps = {
  tags: [],
  addLink: jest.fn(),
  getAllTags: jest.fn(),
}

beforeEach(() => {
  Component = setUp()
})
it('checking prop types', () => {
  let response;
  const component = CreateLink;
  const requiredValues = {};
  response = setPropTypes({component, requiredValues, prop: 'addLink', value:() => {}})
  expect(response).toBeUndefined();
  
  response = setPropTypes({component, requiredValues, prop: 'tags', value: []})
  expect(response).toBeUndefined();
  
});

test('take snapshot', () => {
  expect(Component).toMatchSnapshot();
});

test('get initial tags', () => {
  expect(initialProps.getAllTags).toHaveBeenCalled()
})

test('render elements', () => {
  expect(findByTestAttr(Component, 'inp-title')).toHaveLength(1)
  expect(findByTestAttr(Component, 'inp-url')).toHaveLength(1)
  expect(findByTestAttr(Component, 'btn-send')).toHaveLength(1)
  expect(findByTestAttr(Component, 'cp-tagList')).toHaveLength(1)
  expect(findByTestAttr(Component, 'cp-tagList').prop('tags').length).toBe(0)
})

describe('interacting with imputs', () => {
  test('sending form with invalid data', () => {
    findByTestAttr(Component, 'btn-send').simulate('click');
    expect(initialProps.addLink).not.toHaveBeenCalled();
  })

  test('sending form with invalid title and valid url', () => {
    findByTestAttr(Component, 'btn-send').simulate('click');
    findByTestAttr(Component, 'inp-url').instance().value = 'url test';
    expect(initialProps.addLink).not.toHaveBeenCalled();
  })
  
  test('sending form with valid title and invalid url', () => {
    findByTestAttr(Component, 'inp-title').instance().value = 'title test';
    findByTestAttr(Component, 'btn-send').simulate('click');
    expect(initialProps.addLink).not.toHaveBeenCalled();
  })  

  test('sending form with valid title and invalid url', () => {
    const newLink = {title: 'test title', url: 'test url', tags: [{id : 1}, {id: 2}, {id: 3}]}
    findByTestAttr(Component, 'inp-title').instance().value = newLink.title;
    findByTestAttr(Component, 'inp-url').instance().value = newLink.url;
    //onTagsSaved update the internal state it must be wrapper in this helper 'act'
    act(() => {
      findByTestAttr(Component, 'cp-tagList').prop('onTagsSaved')(newLink.tags);
    })

    const getTagId = tags => tags.map(tag => tag.id);
    findByTestAttr(Component, 'btn-send').simulate('click');
    expect(initialProps.addLink).toHaveBeenCalledWith({...newLink, tags: getTagId(newLink.tags)});
    expect(findByTestAttr(Component, 'inp-title').instance().value).toBe('');
    expect(findByTestAttr(Component, 'inp-url').instance().value).toBe('');
  })
})

function setUp(props = {}){
  return mount(<CreateLink {...initialProps} {...props} />)
}
