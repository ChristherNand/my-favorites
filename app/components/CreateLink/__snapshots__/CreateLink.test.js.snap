// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`take snapshot 1`] = `
<div
  className="createLink"
>
  <h2>
    Create Link
  </h2>
  <input
    className="createLink__contInputs__title"
    onFocus={[Function]}
    placeholder="Title"
  />
  <input
    className="createLink__contInputs__url"
    onFocus={[Function]}
    placeholder="Url"
  />
  <button
    className="createLink__send"
    onClick={[Function]}
    type="button"
  >
    Send
  </button>
  <div
    className="createLink__contTags --flex"
  >
    <AutoComplete
      autoHide={false}
      clearAfterSelecting={true}
      onSelected={[Function]}
      placeHolder="Add Tags"
      propertyFilter="name"
    />
    <div
      className="contOptions"
    />
  </div>
</div>
`;
